internal/xds/rbac/matchers.go

```diff
@@ -0,0 +1,404 @@
+/*
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package rbac
+
+import (
+	"fmt"
+	"net"
+	"regexp"
+
+	v3corepb "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	v3rbacpb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	v3route_componentspb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	v3matcherpb "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+	matcher "google.golang.org/grpc/internal/xds/matcher"
+)
+
+// TODO: Do we need errors at every node? Should we have checks on each node (i.e. destinationPort should be within a certain range)
+
+// The rbacMatcher interface will be used by the RBAC Engine to help determine
+// whether incoming RPC requests should be allowed or denied. There will be many
+// types of matchers, each instantiated with part of the policy used to
+// instantiate the RBAC Engine. These constructed matchers will form a logical
+// tree of matchers, which data about any incoming RPC's will be passed through
+// the tree to help make a decision about whether the RPC should be allowed or
+// not.
+type rbacMatcher interface {
+	matches(args *EvaluateArgs) bool
+}
+
+// policyMatcher helps determine whether an incoming RPC call matches a policy.
+// A policy is a logical role (e.g. Service Admin), which is comprised of
+// permissions and principals. A principal is an identity (or identities) for a
+// downstream subject which are assigned the policy (role), and a permission is
+// an action(s) that a principal(s) can take. A policy matches if both a
+// permission and a principal match, which will be determined by the child or
+// permissions and principal matchers.
+type policyMatcher struct {
+	permissions *orMatcher
+	principals  *orMatcher
+}
+
+func newPolicyMatcher(policy *v3rbacpb.Policy) (*policyMatcher, error) {
+	permissions, err := matchersFromPermissions(policy.Permissions)
+	if err != nil {
+		return nil, err
+	}
+	principals, err := matchersFromPrincipals(policy.Principals)
+	if err != nil {
+		return nil, err
+	}
+	return &policyMatcher{
+		permissions: &orMatcher{
+			matchers: permissions,
+		},
+		principals: &orMatcher{
+			matchers: principals,
+		},
+	}, nil
```

Comment:
Really, does this not read better?

```golang
	return &policyMatcher{
		permissions: &orMatcher{matchers: permissions},
		principals: &orMatcher{matchers: principals},
	}, nil
```

If you don't, you dont have to feel compelled to change.